### ISSUE TITLE
Training not converging using the provided code and script

### ISSUE THREAD

**Issue by Ellenisawake** [liyunsheng13/DRT#5]
Hi yunsheng, thanks a lot for sharing this implementation.
I had an issue when trying to reproduce your result on the DomainNet dataset. Using the given script for clipart domain and using the pre-trained ResNet-101 model, the training cannot converge after 15 epochs (as mentioned in the paper on page 5). I plotted the loss curve as following, which shows that the total loss (loss1 and loss2 as well) does not decrease as expected. The alongside test accuracy also stays at around 0.95%. **Do you have any idea what may have caused this issue? Or are there some parameters to change to get the training going?** Thanks in advance!!
![DRT_clipart](https://user-images.githubusercontent.com/12854625/126809942-240664e6-7c5c-4b0f-a65f-1089912b9779.png)

**Comment by liyunsheng13** [comment]
I'm not sure what this problem is. If you just use the default training script, I think the only reason that might cause this problem is the data. Have you checked the image and its corresponding label are correct?

**Comment by e96031413** [comment]
I trained the DRT using pre-trained ResNet101 for 15 epochs with my own dataset(1 million+ images) and got the following result durning testing. I can sure that the image and its corresponding label are correct.

The loss is negative and the accuracy is almost in 42%~43%

```python
Test set: Average loss: 1.9757, Accuracy: 247.0/6652 (3.71%) (45.81%) (45.81%)
Test set: Average loss: -1.1454, Accuracy: 2881.0/6652 (43.31%) (45.29%) (45.29%)
Test set: Average loss: -1.3885, Accuracy: 2853.0/6652 (42.89%) (46.09%) (46.09%)
Test set: Average loss: -1.0832, Accuracy: 2853.0/6652 (42.89%) (46.09%) (46.09%)
Test set: Average loss: -1.3266, Accuracy: 2916.0/6652 (43.84%) (46.09%) (46.09%)
Test set: Average loss: -1.4154, Accuracy: 2896.0/6652 (43.54%) (45.66%) (45.66%)
Test set: Average loss: -1.5388, Accuracy: 2878.0/6652 (43.27%) (45.13%) (45.13%)
Test set: Average loss: -1.2153, Accuracy: 2797.0/6652 (42.05%) (45.97%) (45.97%)
Test set: Average loss: -1.3479, Accuracy: 2798.0/6652 (42.06%) (43.64%) (43.64%)
Test set: Average loss: -1.2534, Accuracy: 2850.0/6652 (42.84%) (46.11%) (46.59%)
Test set: Average loss: -1.1385, Accuracy: 2839.0/6652 (42.68%) (45.94%) (45.91%)
Test set: Average loss: -1.1853, Accuracy: 2853.0/6652 (42.89%) (45.91%) (45.90%)
Test set: Average loss: -1.3088, Accuracy: 2853.0/6652 (42.89%) (45.94%) (45.93%)
Test set: Average loss: -1.4248, Accuracy: 2853.0/6652 (42.89%) (46.06%) (46.06%)
Test set: Average loss: -1.6749, Accuracy: 2853.0/6652 (42.89%) (46.06%) (46.08%)
```

**Comment by liyunsheng13** [comment]
Are the results the same as your expectation?

**Comment by e96031413** [comment]
Hi @liyunsheng13,

I haven't encountered the negative loss before. But according to the Test set loss, it seems that the model converges.

I don't know whether the situation is related to my dataset. 

My dataset contains 1.3 million images, 35 domains. There are 6 labels for each domain. (e.g. label 0, label 1, label 2, ...... , label 5) But not every domain contains all the 6 labels, some might contain part of the labels (e.g. Domain A might contains label 0 and label 1 only; while some contain all the 6 label)

I use 34 domains as source, and randomly select 1 domains as target.

My expectation is that the model can distinguish the target domain, and get 80%~90% test accuracy.

**Comment by liyunsheng13** [comment]
The paper is for mutli-source unsupervised domain adaptation problem. It cannot provide a similar performance as the supervised case. I have no idea about the dataset you used but based on its complexity you described I think the results given by DRT (~40%) make sense to me.

**Comment by e96031413** [comment]
@liyunsheng13 Thanks for your reply, I can catch the idea of your explanation.

I appreciate your help very much.

**Comment by Ellenisawake** [comment]
> I'm not sure what this problem is. If you just use the default training script, I think the only reason that might cause this problem is the data. Have you checked the image and its corresponding label are correct?

Hi Yunsheng, many thanks for your reply.
Yes I can confirm that the images and corresponding labels are correct.  
One thing to confirm:  in each training, the **trg_path** points to the **train split** of one domain in DomainNet, the **src_path** points to the **train splits** of the **rest of the 5 domains**, and the **val_path** points to the **test split** of the target domain. Is this correct?

**Comment by liyunsheng13** [comment]
Yes.

**Comment by Ellenisawake** [comment]
I see... I am using PyTorch v1.7, would this cause any issue?

**Comment by liyunsheng13** [comment]
I think the version of pytorch should not be a problem. Did you make any changes to the code?

**Comment by Ellenisawake** [comment]
Found the problem where it went wrong... It was the data loader which wasn't properly indexed. My stupid mistake. The training is now up and running normal with decent test performance. I'll close this. Thanks for the help! 😁

**Comment by liyunsheng13** [comment]
Great! Hope the code can help your research.
